yt_dlp/networking/common.py

```diff
@@ -0,0 +1,476 @@
+    def extensions(self):
+        return self._extensions
+
+    @extensions.setter
+    def extensions(self, extensions):
+        if not isinstance(extensions, dict):
+            raise TypeError('extensions must be of type dict')
+        self._extensions = extensions
+
+    @data.setter
+    def data(self, data: _TYPE_REQ_DATA):
+        # Try catch some common mistakes
+        if data is not None and (
+            not isinstance(data, (bytes, io.IOBase, Iterable)) or isinstance(data, (str, Mapping))
+        ):
+            raise TypeError('data must be bytes, iterable of bytes, or a file-like object')
+
+        if data == self._data and self._data is None:
+            self.headers.pop('Content-Length', None)
+
+        # https://docs.python.org/3/library/urllib.request.html#urllib.request.Request.data
+        if data != self._data:
+            if self._data is not None:
+                self.headers.pop('Content-Length', None)
+            self._data = data
+
+        if self._data is None:
+            self.headers.pop('Content-Type', None)
+
+        if 'Content-Type' not in self.headers and self._data is not None:
+            self.headers['Content-Type'] = 'application/x-www-form-urlencoded'
+
+    @property
+    def headers(self) -> CaseInsensitiveDict:
+        return self._headers
+
+    @headers.setter
+    def headers(self, new_headers: Mapping):
+        """Replaces headers of the request. If not a CaseInsensitiveDict, it will be converted to one."""
+        if isinstance(new_headers, CaseInsensitiveDict):
+            self._headers = new_headers
+        elif isinstance(new_headers, Mapping):
+            self._headers = CaseInsensitiveDict(new_headers)
+        else:
+            raise TypeError('headers must be a mapping')
+
+    def update(self, url=None, data=None, headers=None, query=None):
+        self.data = data or self.data
+        self.headers.update(headers or {})
+        self.url = update_url_query(url or self.url, query or {})
+
+    def copy(self):
+        return self.__class__(
+            url=self.url,
+            headers=copy.deepcopy(self.headers),
+            proxies=copy.deepcopy(self.proxies),
+            data=self._data,
+            extensions=copy.copy(self.extensions),
+            method=self._method,
+        )
+
+
+HEADRequest = functools.partial(Request, method='HEAD')
+PUTRequest = functools.partial(Request, method='PUT')
+
+
+class Response(io.IOBase, AutoCloseMixin):
+    """
+    Abstract base class for HTTP response adapters.
+
+    Interface partially backwards-compatible with addinfourl and http.client.HTTPResponse.
+
+    @param raw: Original response.
+    @param url: URL that this is a response of.
+    @param headers: response headers.
+    @param status: Response HTTP status code. Default is 200 OK.
+    @param reason: HTTP status reason. Will use built-in reasons based on status code if not provided.
+    """
+
+    def __init__(
+            self,
+            raw,
```

Comment:
change to `fp` and type as a file maybe to make it more clear.

If an implementation does not act like a file the methods will need to add be adapted anyways.